velox/functions/prestosql/aggregates/tests/PrestoHasherTest.cpp

```diff
@@ -0,0 +1,261 @@
+}
+
+TEST_F(PrestoHasherTest, floatType) {
+  assertHash<float>(
+      {1.0f, 0.0f, std::nanf("0"), 99999.99f},
+      {-6641611864725600567L, 0, 6018425597037803642, -6781949276808836923});
+}
+
+TEST_F(PrestoHasherTest, VarcharType) {
+  assertHash<StringView>(
+      {StringView("abcd"),
+       StringView(""),
+       std::nullopt,
+       StringView(u8"Thanks \u0020\u007F")},
+      {-2449070131962342708, -1205034819632174695, 0, 2911531567394159200});
+}
+
+TEST_F(PrestoHasherTest, boolType) {
+  assertHash<bool>({true, false}, {1231, 1237});
+}
+
+TEST_F(PrestoHasherTest, arrays) {
+  auto O = [](std::vector<std::optional<int64_t>> data) {
+    return std::make_optional(data);
+  };
+
+  auto baseArrayVector = vectorMaker_.arrayVectorNullable<int64_t>(
+      {O({1, 2}),
+       O({3, 4}),
+       O({4, 5}),
+       O({6, 7}),
+       O({8, 9}),
+       O({10, 11}),
+       O({12, std::nullopt})});
+
+  auto arrayExpected = std::vector<int64_t>(
+      {4329740752828760473,
+       655643799837771513,
+       8633635089947142034,
+       9138382565482297209,
+       1065928229506940121,
+       8616704993676952121,
+       1942639070761256766});
+
+  assertHash(arrayExpected, baseArrayVector);
+
+  // Array of arrays.
+  auto arrayOfArrayVector =
+      vectorMaker_.arrayVector({0, 2, 4}, baseArrayVector);
+
+  auto arrayofArraysExpected = std::vector<int64_t>(
+      {5750398621562484864, 79909248200426023, 2145437271247615678});
+
+  assertHash(arrayofArraysExpected, arrayOfArrayVector);
+
+  // Array with nulls
+
+  std::vector<std::optional<std::vector<std::optional<int64_t>>>> data = {
+      O({std::nullopt}),
+      O({1, 2, 3}),
+      O({1024, std::nullopt, -99, -999}),
+      O({}),
+      O({std::nullopt, -1}),
+  };
+
+  auto arrayWithNulls = vectorMaker_.arrayVectorNullable(data);
+
+  auto arrayWithNullsExpected = std::vector<int64_t>{
+      0, -2582109863103049084, 2242047241851842487, 0, -6507640756101998425};
+
+  assertHash(arrayWithNullsExpected, arrayWithNulls);
+}
+
+TEST_F(PrestoHasherTest, maps) {
+  auto mapVector = makeMapVector<int64_t, int64_t>(
+      3,
+      [](auto /**/) { return 1; },
+      [](auto row) { return row; },
+      [](auto row) { return row + 1; });
+
+  auto mapExpected = std::vector<int64_t>{
+      9155312661752487122, -6461599496541202183, 5488675304642487510};
+
+  assertHash(mapExpected, mapVector);
+
+  auto mapOfArrays =
+      makeMapofArraysVector({1, 2, 3}, {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}});
+  assertHash(
+      {-6691024575166067114, -7912800814947937532, -5636922976001735986},
+      mapOfArrays);
+}
+
+TEST_F(PrestoHasherTest, rows) {
+  auto firstVector = makeFlatVector<int64_t>({1, 3});
+  auto secondVector = makeFlatVector<int64_t>({2, 4});
+
+  auto row = makeRowVector({firstVector, secondVector});
+
+  auto rowExpected =
+      std::vector<int64_t>{4329740752828761434, 655643799837772474};
+
+  assertHash(rowExpected, row);
+}
+
+TEST_F(PrestoHasherTest, dictionary) {
```

Comment:
It would be nice to extend assertHash to add testing of encodings, e.g. wrap input in a dictionary, compute hashes, assert result. Then, wrap input in a constant.

Reply:
Done.